Implement exception groups

ZeroIntensity opened the pull request:
# What does this PR do?

This implements `except*` as outlined in PEP 654.

Note that this is currently a very rough outline of how this is going to work, and definitely not finished yet. The `except*` reformulation is going to be the following:

```py
try:
    ...
except BaseException:  # Equivalent to except* exception_types where exception_types is a tuple
    _tmp_error = sys.exc_info()[1]
    if isinstance(_tmp_error, BaseExceptionGroup):
        err = _tmp_error
    else:
        err = ExceptionGroup('', (_tmp_error,))

    for _type in exception_types:
        for _val in err.exceptions:
            if isinstance(_val, _type):
                break
        else:
            raise
```

# Why was it initiated? Any relevant Issues?

See #2386.

# PR Checklist

- [x] Correct base branch selected? Should be `develop` branch.
- [ ] Enabled commit hook or executed `./bin/autoformat-nuitka-source`.
- [ ] All tests still pass. Check the Developer Manual about `Running the Tests`. There are GitHub
  Actions tests that cover the most important things however, and you are welcome to rely on those,
  but they might not cover enough.
- [ ] Ideally new features or fixed regressions ought to be covered via new tests.
- [ ] Ideally new or changed features have documentation updates.

## Summary by Sourcery

Implement core PEP 654 exception group support by extending exception matching, AST transformations, and code generation to handle 'except*' constructs.

New Features:
- Add preliminary support for PEP 654 exception groups and 'except*' handlers in try/except statements.

Enhancements:
- Extend the C API with EXCEPTION_GROUP_MATCH_BOOL to enforce correct exception group matching semantics.
- Introduce ExpressionComparisonExceptionGroupMatch and integrate the 'exception_group_match' comparator into code generation.
- Augment try/except AST reformulation to emit specialized handling logic for starred exception handlers.

dcharbon committed:
Standalone: Add support for "xgboost.core" on macOS as well

kayhayen committed:
UI: Fix, add back mnemonic, it's used to allow to disable the warning

* Tests and distutils integration are using it. For the later
  it doesn't really apply, since that's in a temporary folder
  and the result will be copied elsewhere anyway.

kayhayen committed:
UI: When using package mode, don't warn about no other follows

kayhayen committed:
Module: Fix, for namespace packages the stub generation failed

* There is no source code to work on in that case.

* Also moved pyi file generation to separate function as a cleanup.

kayhayen committed:
Tests: Added support for "NUITKA_EXTRA_OPTIONS" to distutils cases with pyproject

* For the other ones, we still need to figure out how we want to do that.

kayhayen committed:
Python3.13: Fix, extension modules "x.x" failed to load when applying the workaround needed for unexposed API

* We added the module to the dictionary before we scanned it, so it appeared in the list of modules to correct wrongly

* Also added more traces and enhanced them for when the hack is used.

kayhayen committed:
Distutils: Use "--module=package" where it makes sense

kayhayen committed:
Standalone: Fix, need to take reference for extension modules returned in "exec_module"

kayhayen committed:
Standalone: Added support for "duckdb"

kayhayen committed:
New pre-release.

kayhayen committed:
New stable release.

kayhayen committed:
Merge branch 'release/2.7'

kayhayen committed:
Merge tag '2.7' into develop

kayhayen committed:
New release cycle.

kayhayen committed:
Fix, do not crash when "--include-module" is used on a built-in module

* Instead warn about these, as that should be unusual.

* Also added TODO about adding builtin modules as a module node type and to keep track of them too.

kayhayen committed:
Optimization: Detect raises with compile time constant values that are not exceptions

* Previously it was crashing if given code like
  "raise NotImplemented" which is wrong to do,
  but shouldn't crash Nuitka of course.

kayhayen committed:
Optimization: Avoid compile constant conditional statements for "assert False" and the like

* We can just do it during the tree building, and
  avoid the noise in optimization to handle it
  there in two steps.

* For known false values we cannot ignore the
  message, as with assignment expressions it
  might introduce a local variable that otherwise
  becomes non-local.

kayhayen committed:
Debugging: Avoid handing out wrong values by accident

kayhayen committed:
Python3.5: Restore stubgen syntax compatibility

kayhayen committed:
Module: Fix, "code" constants blob mode didn't work with C++ fallback

* This affects only very old gcc on Linux really,
  but it's still a valid fix, we already covered
  "linker" in this way.

kayhayen committed:
Fix, need to avoid warning about unused variable

* Not just check this in debug builds, abort if it fails in normal mode too.

kayhayen committed:
Anti-Bloat: Avoid "docutils" usage in "markdown2" package

kayhayen committed:
Standalone: Added support for "nicegui" package

kayhayen committed:
macOS: Workaround for data file names that cannot be signed in app bundles

* We create top level symlinks, such that these can be signed instead, and they are legal.

kayhayen committed:
fixup! Bump copyright year

kayhayen committed:
fixup! AIX: Added working code for determining running binary path

kayhayen committed:
Plugins: Fix, do not crash when importing "pkg_resources" gives "PermissionError"

* This happens with installation with improper permissions, and then Nuitka crashes where Python doesn't (yet).

kayhayen committed:
AIX: Added support for scanning rpaths from COFF binaries

kayhayen committed:
Anti-Bloat: Better handling of "astropy" package

kayhayen committed:
Watch: Added TODO for container pipenv usage

kayhayen committed:
Debugging: Enhance exception to output mismatching module name

* This actually happened in CI that is not easy to externally reproduce.

kayhayen committed:
Python3.12.0: Fix, cannot use unknown identifier

* This wasn't added before 3.12.1 and therefore
  won't compile with the 3.12.0 release

kayhayen committed:
Fix, for some setups extension module suffix order can trigger assertions

* The derived module name could be calculated
  using the wrong suffix.

* This triggered assertions with Python2 on
  CentOS7 at least.

* Also clean up to make it possible to enable
  logging per module search call.

kayhayen committed:
Linux: Fix, do not use the CRC32 helpers as "extern"

* Otherwise they might be confused with the zlib header ones.

kayhayen committed:
Python2.6: Fix, didn't have "sys.version_info.releaselevel"

* Added helper for accessing this, that is more portable,
  and that also cleans up things somewhat.

kayhayen committed:
Debugging: Fix, wrong assertion

* This also lead to a warning with older gcc
  versions.

kayhayen committed:
Python2.6: Fix, this method is now used too

kayhayen committed:
Fix, some "orderedset" implements do not have "union" method

* Also, update is what the code intends to do, so this is a cleanup as well

kayhayen committed:
Python2.6: Fix, "staticmethod" objects didn't have "__func__"

* We therefore need to remember the original function.

kayhayen committed:
Visual Code: Added Python2.6 config for Win32 to use for comparisons

kayhayen committed:
Plugins: Avoid hard coding preload module names, these should be dynamic

* Currently they are not, but we will fix that.

kayhayen committed:
fixup! Plugins: Add support for "cloudpickle"

kayhayen committed:
Module: The "__compiled__.main" value didn't reflect the actual module

* When loading a separately compiled module into a package
  as sub-package, the package context was not correctly
  derived.

* This makes "dill-compat" usable with modules/packages
  loaded as sub-packages.

kayhayen committed:
New hotfix release.

kayhayen committed:
Python2.6: We now use this outside of asserts too

* Strangely only some versions of Python2.6 don't have
  it.

kayhayen committed:
Standalone: Added needed data files for "ursina"

* This is given a warning if it's not included, probably a lot more should be or even less, only one model might be required.

kayhayen committed:
Merge branch 'hotfix/2.7.1'

kayhayen committed:
Merge tag '2.7.1' into develop

kayhayen committed:
New pre-release.

kayhayen committed:
Plugins: Avoid using builtins dictionary with loader type to register with "pkg_resources"

* Instead we can just use the "__loader__" value of the
  registering module, that is what we are looking for.

* Not polluting the builtins dict has to be preferred
  always.

kayhayen committed:
Linux: Avoid name used in "libz"

* This can cause link time errors apparently, renaming it
  should avoid it, also make it static, as we don't need
  it elsewhere.

kayhayen committed:
fixup! Plugins: Fix, do not crash when importing "pkg_resources" gives "PermissionError"

kayhayen committed:
macOS: Fixed issue with framwork resources for QtWebKit

* This special handling is no longer necessary since it has
  become the default for macOS anyway to symlink files.

* This also removes a duplication of said files that probably
  should have been avoided all along by using symlinks.

kayhayen committed:
Anaconda: Follow changes in newer conda packages for PySide2

* More paths mismatch the configured paths and need to be
  adapted, and previous guesses are no longer correct in
  some cases.

* Also normalize paths properly with the use of "getNormalizedPath"
  which might be good for MSYS2 and its always a cleanup.

kayhayen committed:
Linux: Fix, was crashing when attempting to modify rpath of statically linked executables

* Some PyPI libraries like "imageio-ffmpeg" have these.

* The implementation of "isStaticallyLinked" is imcomplete
  in terms of platform supported, so we can't really use
  it for much else yet.

kayhayen committed:
Python3.14: Use new API for accessing path config

kayhayen committed:
Python Build Standalone: Added built-in modules for debug build to not raise

* This adds the modules are compiled into this flavor
  which we also could have seen elsewhere, but so far
  did not.

kayhayen committed:
New hotfix release.

kayhayen committed:
Minor spelling cleanups

kayhayen committed:
fixup! Plugins: Add support for "cloudpickle"

kayhayen committed:
Merge branch 'hotfix/2.7.2'

kayhayen committed:
Merge tag '2.7.2' into develop

kayhayen committed:
New pre-release.

kayhayen committed:
amend! Anaconda: Follow changes in newer conda packages for PySide2

Anaconda: Follow changes in newer conda packages for PySide6

* More paths mismatch the configured paths and need to be
  adapted, and previous guesses are no longer correct in
  some cases.

* Also normalize paths properly with the use of "getNormalizedPath"
  which might be good for MSYS2 and its always a cleanup.

kayhayen committed:
amend! Anaconda: Follow changes in newer conda packages for PySide2

Anaconda: Follow changes in newer conda packages for PySide6

* More paths mismatch the configured paths and need to be
  adapted, and previous guesses are no longer correct in
  some cases.

* Also normalize paths properly with the use of "getNormalizedPath"
  which might be good for MSYS2 and its always a cleanup.

kayhayen committed:
macOS: Fix, we should provide the "apple" target

* In some configurations, the linker can otherwise
  warn about using "unknown" in combination with
  other things.

kayhayen committed:
macOS: Disable static libpython for pyenv on macOS

* It's heavily broken and unlikely to be repaired ever.

kayhayen committed:
Standalone: Fix, scanning for standard library in PATH should not be necessary

* For Anaconda Python, we already do solve using a
  reduced one with the necessary stuff.

* This was left mostly because it requires a proper
  view of inside paths for all flavors. This has
  been added for Homebrew and macOS CPython now.

kayhayen committed:
macOS: Catch error of not providing description for "--macos-app-protected-resource" option

kayhayen committed:
Plugins: Check for Qt bindings module presence without insisting on metadata

* In some configurations, there is no metadata, as has
  been observed recently with Homebrew and PySide6 from
  their bottle. With "uv" it appears to maybe happening
  as well. Making metadata mandatory for Nuitka only is
  not good anyway.

* All we mean to do there is to really enforce that they
  are ordered by shorted name, should the metadata be looked
  up, which it is for reporting purposes at least.

* Added new plugin entry point that is called right before
  compilation and where the check can be done if the binding
  module is located.

* This solves an older TODO about not doing things during the
  plugin initialization.

kayhayen committed:
Fix, for PySide6 slots in base classes could remain unhandled

* As a result, only their first using class worked, not any
  subsequent ones.

kayhayen committed:
fixup! Fix, need to prevent Anaconda and poetry without virtualenv being used in a mixed fashion

kayhayen committed:
New hotfix release.

kayhayen committed:
Standalone: Added support for pypylon package

kayhayen committed:
Windows: For onefile DLL mode, all environment variables need to be set for QtWebEngine

* Defaults of Qt point to near the onefile binary,
  which is now always wrong.

* Added TODO to do this for all platforms, since
  using default scan results is probably not a good
  a good idea anyway.

kayhayen committed:
Plugins: Avoid including debug binary for Qt Webkit

kayhayen committed:
fixup! Debugging: Remove wrong assertions, handle NULL value for published exception printing

kayhayen committed:
Standalone: Make it clear that "torch" uses DLLs from "nvidia" package

kayhayen committed:
fixup! Python3.12.0: Fix, cannot use unknown identifier

kayhayen committed:
Merge branch 'hotfix/2.7.3'

kayhayen committed:
Merge tag '2.7.3' into develop

kayhayen committed:
Plugins: List available plugin families if "--include-qt-plugin" cannot find one.

kayhayen committed:
UI: Remove obsolete mode value options

* These are only kept for compatibility

kayhayen committed:
UI: Make "--disable-plugin" visible on Nuitka-Action too

* Sometimes there are uses for this, don't be as opiniated.

kayhayen committed:
Standalone: More data directories for "ursina" and "panda3d" package

kayhayen committed:
fixup! Windows: Use icons given for Windows automatically with PySide6

kayhayen committed:
Python3.10+: Fix, the matched value was not released and leaked memory

kayhayen committed:
AIX: Added "dladdr" implementation needed

* This should enable us to use module mode on AIX.

kayhayen committed:
Fix, need to annotate exception exit for larger ranges that we do not optimize away

csd4ni3l committed:
Standalone: Added "arcade" gl backends

kayhayen committed:
macOS: Make sure SSL certificates are available

* We can use the pip ones or certifi package if that's not available.

* This avoids breaking on systems where the certificates are not installed

kayhayen committed:
UI: Better warning text for macOS icons

kayhayen committed:
Minor cleanup

* Can just check for app bundle only

kayhayen committed:
macOS: Do not sign through symlinks

* This also avoid using them unless absolutely necessary
  and then only for the parts that absolutely are required
  to have them, since they can cause issues apparently.

kayhayen committed:
UI: Do not use deprecated option for app bundles in options nanny plugin

kayhayen committed:
Standalone: Added data files of "pygameextra"

kayhayen committed:
Standalone: Fix, for qml plugins on "macOS" with newer PySide6

kayhayen committed:
Plugins: Make sure to use macOS target architecture

* Otherwise compiling for x86_64 using "arch -x86_64 python" for PySide2 crashed with standard CPython universal binaries

kayhayen committed:
UI: When a forbidden import comes from an implicit import, report that properly

* Sometimes ".pyi" files from extension modules cause an import, but it wasn't clear which one.

kayhayen committed:
Standalone: Added support for newer "pydantic"

kayhayen committed:
Standalone: Added support for "libonnxruntime" on macOS too

kayhayen committed:
macOS: Enhanced PySide2 support

* We no longer require onefile mode except for QtWebEngine which segfaults if we don't

kayhayen committed:
Tests: Do not fail on rejected download prompts

kayhayen committed:
Minor spelling cleanup

kayhayen committed:
UI: Fix, was crashed on rejected "ccache" download on macOS

kayhayen committed:
Quality: Minor pylint cleanup

kayhayen committed:
Debugging: Also find onefile temporary files using long paths

kayhayen committed:
Windows: Fix, need to shorten paths used in onefile dll mode too

* This affects onefile DLL and might cause issues
  for users with unicode paths, that we want to
  avoid.

* Also affects module mode, which may also benefit
  somewhat from the change.

kayhayen committed:
fixup! Python Build Standalone: Added built-in modules for debug build to not raise

kayhayen committed:
Merge branch 'hotfix/2.7.4'

kayhayen committed:
Merge tag '2.7.4' into develop

kayhayen committed:
Release: Handle newer setuptools properly

kayhayen committed:
Minor cleanup

kayhayen committed:
UI: Warn about compiling "__main__.py" which should be avoided

kayhayen committed:
Plugins: Enhanced workarounds for PySide6

* This also covers single shot timer callbacks

kayhayen committed:
Plugins: Include binding module in accelerated mode

kayhayen committed:
Tests: Cleanup, avoid using "doctests"

* This is for using this in standalone mode for reference counting tests without a warning

kayhayen committed:
Tests: Change test to be robust against short paths for package containing directory

kayhayen committed:
New hotfix release.

kayhayen committed:
Merge branch 'hotfix/2.7.4'

kayhayen committed:
Merge tag '2.7.4' into develop

kayhayen committed:
New pre-release.

kayhayen committed:
Onefile: Fix, splash screen was showing in DLL mode twice or more

* Also it couldn't be stopped.

kayhayen committed:
Onefile: The DLL mode didn't work with MSVC 14.2 or earlier

kayhayen committed:
New hotfix release.

kayhayen committed:
Anti-Bloat: Avoid using "unittest" in "reportlab" package

* This adds a new feature of "global_change_function" to the
  anti-bloat engine, to address all "test" functions in all
  sub-modules at once.

kayhayen committed:
Debugging: Make PySide6 plugin enforce "--no-debug-immortal-assumptions" when "--debug" is on

kayhayen committed:
Standalone: Added support for newer "skimage" package

kayhayen committed:
Standalone: Enhanced config for "pypylon" on macOS

* This is not working yet, we need to add something new and specific for "frameworks" as dependencies.

kayhayen committed:
Merge branch 'hotfix/2.7.5'

kayhayen committed:
Merge tag '2.7.5' into develop

kayhayen committed:
Quality: Avoid writing formatted files with same contents

* That avoids stirring up tools that listen to changes.

* The Nuitka website autobuilder otherwise rebuilds per
  release post on docs update.

kayhayen committed:
Standalone: Added support for "ansicon" on Windows

* Missing DLL added

kayhayen committed:
Standalone: Added support for "jinxed" package

* The terminal information module is dynamically loaded.

kayhayen committed:
Anti-Bloat: Also avoid "setuptools" in "jax.version"

kayhayen committed:
macOS: Add ability to unlock keyring files with password

* This makes code signing directly possible.

kayhayen committed:
UI: Only check onefile target spec if it's actually onefile

kayhayen committed:
Cleanup, also avoid "UserWarning" given by "pkg_resources" module

kayhayen committed:
macOS: Workaround for symlinks didn't work with "--output-directory"

* Need of course not make assumptions about dist directory only
  having a single slash before the data file paths.

kayhayen committed:
Anti-Bloat: Remove its testing framework when using "langsmith" package

kayhayen committed:
Standalone: Added support for "paddlex" package

kayhayen committed:
macOS: Added support for tk-inter with 3.13 and CPython official

* Need to follow changes to use framework files for Tcl and Tk.

timminator committed:
Standalone: Add support for 'PyTaskbar' package

kayhayen committed:
New hotfix release.

kayhayen committed:
fixup! Release: Handle newer setuptools properly

kayhayen committed:
Fix, data files included were no longer checked against DLLs included by program

* We had changed the order of data file copying and DLL copying
  and as a result, the information of DLLs to conflict with was
  incomplete.

* Moved signing for macOS to be a dedicated third step that gets
  information from both phases instead and reverted to old order.

kayhayen committed:
UI: Output the relevant part in error message for data dir options

kayhayen committed:
UI: Corrected wrong punctuation in error message

kayhayen committed:
Merge branch 'hotfix/2.7.6'

kayhayen committed:
Merge tag '2.7.6' into develop

kayhayen committed:
New pre-release.

kayhayen committed:
Doc: Man page examples didn't do standalone and onefile

* Also use new style options

kayhayen committed:
UI: Allow specifying the dist folder name

* Also unified how filenames are checked, and make empty path illegal too for where we check general paths.

kayhayen committed:
Doc: Fix, man pages were not rendered correctly

* There is an issue with calling "--version --help" where the later
  takes precedence, avoid that by knowing the context we run in, as
  otherwise we had options duplicated.

* We also created the wrong kind of command line to ask for help, which
  no longer outputs full help for a while, so "nuitka-run" was almost
  empty.

kayhayen committed:
AIX: Another instance of using "$" in identifiers fixed

kayhayen committed:
Scons: Fix, avoid define of target OS on non-Windows

kayhayen committed:
AIX: Define target arch towards the gcc compiler

kayhayen committed:
Scons: Cleanup, refactor to use functions for OS specific target architecture options

kayhayen committed:
AIX: This function is not available in DLL mode

* We cannot get the original argv as easily, however, we could use the current one if need be.

kayhayen committed:
AIX: Remove dead code, this never worked

kayhayen committed:
AIX: For module mode, also need to link against libpython on that OS

kayhayen committed:
AIX: Don't try to update rpath, not supported

kayhayen committed:
Standalone: Don't assume that patchelf is Linux only

* Also this avoids the error given on AIX when it
  tries to set the RPATH but cannot use patchelf
  on a COFF system.

kayhayen committed:
AIX: Avoid python installation folder in compiled program library path

kayhayen committed:
macOS: Added option to disallow an app to run in multiple instances

kayhayen committed:
macOS: Added new option to manpage

kayhayen committed:
Anti-Bloat: Avoid compiling generated pb2 files

b-phi committed:
Release: Add extra dependency group for Nuitka build-backend

* This is for use in dependencies of "build-system" in the
  "pyproject.toml" files.

* Using "build-wheel" for now, future releases will expand
  on this.

kayhayen committed:
User Manual: Updated to use "Nuitka[build-wheel]" and also updated a few obsolete uses

kayhayen committed:
Release: Add more extra dependency groups for Nuitka

* The dependency on "zstandard" will in the future only be
  on "Nuitka[onefile]" and be removed from "Nuitka" to
  avoid cluttering builds that don't need it.

kayhayen committed:
Release: Add a "all" dependency group for simplicity

kayhayen committed:
GitHub: Use "--mode" in issue template as well

kayhayen committed:
Scons: Cleanup, the debug mode is handled in common setup already

* This was still repeating it on the onefile config

kayhayen committed:
Scons: Do not disable all warnings with clang

* This lived here unrecognized for some time.

kayhayen committed:
Scons: Recognize warnings causing build error and point out the workaround

* Also abort build immediately after one file failed, no
  need to continue with the others.

* For this we need to know the processes to terminate and
  keep the threads in a list.

* This is only non-Windows for now.

kayhayen committed:
Python3.11: Fix, for this and 3.12 we must not expose descriptors for compiled methods

* With 3.13 they came back and the issue that was tried to
  address must have been solved elsewhere, since tests across
  all versions work with this change.

kayhayen committed:
Standalone: Added support for newer "numpy"

kayhayen committed:
UI: Added support for using "rich" progress bar for downloads

* This pioneers the use of "rich" in Nuitka, the other progress bars will follow too.

* The "rich" is preferred from the pip vendored copy and as such likely to be present.

* The "tqdm" is only used if "rich" cannot be found.

kayhayen committed:
Minor spelling cleanups

* Avoid using "ifdef" in templates, lets just avoid it generally.

* Added missing slot function names to the ignored word list.

* Renamed variables rslot, lslot, to be more verbose, that's
  for better understanding of their use anyway.

kayhayen committed:
GitHub: Enhanced wordings for bug report template

* This is in hopes to achieve more compliance to it.

kayhayen committed:
Reports: For Python3.13, add the GIL usage or not

kayhayen committed:
New hotfix release.

kayhayen committed:
Merge branch 'hotfix/2.7.7'

kayhayen committed:
Merge tag '2.7.7' into develop

kayhayen committed:
New pre-release.

kayhayen committed:
fixup! UI: Added support for using "rich" progress bar for downloads

kayhayen committed:
fixup! Scons: Do not disable all warnings with clang

kayhayen committed:
Scons: Recognize warnings from clang on macOS too

kayhayen committed:
GitHub: Ask for our own metadata tool output

kayhayen committed:
Standalone: Added support for older "vtk" package

kayhayen committed:
UI: Sort --list-distribution-metadata output and remove duplicates

kayhayen committed:
Python3.13.4: Reject broken CPython official release for Windows

* The link library included is not the one needed for GIL, and as such it breaks Nuitka heavily and must be errored out on

kayhayen committed:
New pre-release.

ZeroIntensity committed:
Rough outline for implementing this.

ZeroIntensity committed:
Implement call node for casting the exception.

ZeroIntensity committed:
Finish reformulation (I think)

ZeroIntensity committed:
Implement some C code.

kayhayen commented on nuitka/build/include/nuitka/exceptions.h:
```diff
@@ -1272,6 +1272,43 @@ NUITKA_MAY_BE_UNUSED static inline int EXCEPTION_MATCH_BOOL(PyThreadState *tstat
+    CHECK_OBJECT(exceptions_got);
+    CHECK_OBJECT(exceptions_checked);
+
+#if PYTHON_VERSION >= 0x3b0
```
The define should apply to the whole function. I already added a file `exception_groups.h`  that has that define itself, but is also included like this:

```
#if PYTHON_VERSION >= 0x3b0
#include "nuitka/exception_groups.h"
#endif
```

So move it to there, and maybe build off the code I already have in that file.

kayhayen commented:
Looking at the diff, one thing you ought to do, is to have the `misc/install-git-hooks.py` executed, so it will autoformat on commit, which will apply a combination of black, isort, clang-format to the code.

I believe the change of the published exception, the place that `sys.exc_info()` is pulling from, is not yet there.

kayhayen commented:
Executing `./bin/autoformat-nuitka-source` will format the code manually as needed too.

kayhayen committed:
Merge branch 'hotfix/2.7.8'

kayhayen committed:
Merge tag '2.7.8' into develop

kayhayen committed:
New pre-release.

kayhayen committed:
Plugins: Avoid loading modules when loading data

* In cases where the file doesn't exist, this loaded the module unnecessarily.

* In case of broken installs that then can crash Nuitka and generally ought to not happen anyway.

kayhayen committed:
Azure: Stop using it, we are going to use GitHub instead

ZeroIntensity committed:
Some refactoring.

ZeroIntensity committed:
Fix failing build.

ZeroIntensity committed:
Fix validation.

ZeroIntensity committed:
Fix matching.

ZeroIntensity committed:
Patch the return value a bit.

ZeroIntensity committed:
Change return value.

ZeroIntensity commented:
Definitely getting closer to fruition. The two main part that I need to iron out is the partial matching. It's not yet clear to me how to execute the `except*` block, but then raise an arbitrary exception group object right afterwards.

ZeroIntensity committed:
Housekeeping.

ZeroIntensity committed:
Start implementation of partial matches.

kayhayen committed:
Plugins: Remove obsolete warning for Linux with upx

* We don't do "appimage" anymore for a while now, so its constraints do not matter.

kayhayen committed:
Scons: Use "c23_embed" for ClangCL too

* These modes are not applied for onefile, which prevents it from full success.

kayhayen committed:
Scons: Avoid warnings for unknown options with clang

kayhayen committed:
Python3.12: Enhanced importing scan to cover setuptools distutils hack

* Now we can find "distutils" in a folder named "_distutils"
  and pass the information upwards, what it was supposed to be.

* For now the old code is still used for non-distutils and we
  assert in debug mode for correctness only.

kayhayen committed:
UI: Added mode "app-dist" for standalone everywhere except "app" on macOS

kayhayen committed:
Minor spelling cleanups

kayhayen committed:
Deployment: Catch modules that error exit on import, while assumed to work perfectly

* This will give people an indication that the "numpy" module is expected to work, and that maybe just the newest version is not.

kayhayen committed:
Minor spelling cleanups

ZeroIntensity committed:
Revert "Start implementation of partial matches."

This reverts commit feaafad292f6c580ed290fa30e5b2966341d0810.

kayhayen committed:
Plugins: Make "global_change_function" work on qualnames rather than function names

* This makes it such that methods are not confused with functions as easily.

kayhayen committed:
Standalone: Added support for "reportlab.graphics.barcode"

kayhayen committed:
Actions: Disable Python 3.13 on Windows, until it's not 3.13.4

* The CPython issue that was happening breaks Nuitka workflows
  on GitHub until 3.13.5 gets used, which apparently is not the
  case now.

kayhayen committed:
fixup! Scons: Recognize warnings causing build error and point out the workaround

kayhayen committed:
Plugins: Major overhaul of UPX plugin

* Use better compression.

* Hint the user at disabling onefile compression where applicable.

* Output warnings about not compressible files.

* Error exit when upx is not found right after startup.

kayhayen committed:
Standalone: Added support for newer "certifi" package

* Unfortunately we cannot avoid "importlib.resources" anymore

kayhayen committed:
Minor spelling cleanups

kayhayen committed:
Onefile: Fix, "containing_dir" was regressed and pointed to DLL dir on Windows

kayhayen committed:
New hotfix release.

kayhayen committed:
Merge branch 'hotfix/2.7.9'

kayhayen committed:
Merge tag '2.7.9' into develop

kayhayen committed:
New pre-release.

kayhayen committed:
fixup! Onefile: Fix, "containing_dir" was regressed and pointed to DLL dir on Windows

kayhayen committed:
New hotfix release.

kayhayen committed:
Merge branch 'hotfix/2.7.10'

kayhayen committed:
Merge tag '2.7.10' into develop

ZeroIntensity committed:
Outline for the try/finally chains.

ZeroIntensity committed:
Hunk a bunch of things together.

ZeroIntensity committed:
Fix the C helper.

ZeroIntensity committed:
Some fixups.

kayhayen commented on nuitka/nodes/ExceptionNodes.py:
```diff
@@ -334,6 +334,33 @@ def __init__(self, source_ref):
+
+    def __init__(self, caught, catching, source_ref):
+        ExpressionBase.__init__(self, source_ref)
+        self.caught = caught
```
The way you announce children to Nuitka, is wrong here, that's probably all your remaining issues.

kayhayen committed:
Use base node generation for exception group match too

ZeroIntensity committed:
Fix some generation shims.

ZeroIntensity committed:
Some formalities with getErrorExitCode()

ZeroIntensity committed:
Fix temp_type argument.

ZeroIntensity committed:
Unpack match tuple items into variables.

ZeroIntensity committed:
Some patchwork.

ZeroIntensity committed:
Fix nested statement sequences.

ZeroIntensity committed:
Fix incorrect is_match value.

ZeroIntensity committed:
Fix double-wrapping of the exception group.

The C helper wraps it for us, we don't need to do it in the reformulation.

ZeroIntensity committed:
Share 'rest' across handlers.

kayhayen commented on nuitka/tree/ReformulationTryExceptStatements.py:
```diff
@@ -201,43 +201,54 @@ def makeTryExceptSingleHandlerNodeWithPublish(
 
+def fix_nested_sequence(statement_sequence, source_ref):
+    # XXX: This is a hack! It seems that we can't have a StatementsSequence()
+    # nested directly under a StatementsSequence(), so we wrap it in an "if True"
```
There is never a need to have them, that's why they are prevented. There is `mergeStatements` if you want to pass a statements sequence, and a statement to and get a tuple which can be used to create a statements sequence.

ZeroIntensity replied:
Well, there's a need to have them when you factor things out into smaller helpers like I've done. I'll switch to `mergeStatements`, but it's a little inconvenient; couldn't Nuitka just flatten out the statements anyway?

kayhayen committed:
Added re-raise statement

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/Nuitka/Nuitka?pullRequest=3509) <br/>All committers have signed the CLA.

ZeroIntensity commented:
Superseded by #3743
